sdk/servicebus/Microsoft.Azure.ServiceBus/src/Primitives/ConcurrentExpiringSet.cs

```diff
@@ -4,33 +4,45 @@
 namespace Microsoft.Azure.ServiceBus.Primitives
 {
     using System;
+    using System.Collections.Generic;
     using System.Collections.Concurrent;
+    using System.Threading;
     using System.Threading.Tasks;
 
     sealed class ConcurrentExpiringSet<TKey>
     {
         readonly ConcurrentDictionary<TKey, DateTime> dictionary;
+        readonly ICollection<KeyValuePair<TKey, DateTime>> dictionaryAsCollection;
```

Comment:
How does this help?

Reply:
Will answer later

Reply:
ConcurrentDictionary implements the `ICollection<KeyValuePair<TKey, TValue>>` interface implicitly. By having a reference to that we can use the `Remove(KeyValuePair<TKey, TValue> kvp)` method which calls into [`TryRemoveInternal`](https://referencesource.microsoft.com/#mscorlib/system/Collections/Concurrent/ConcurrentDictionary.cs,1453) with `matchValue` set to `true` which then makes sure that the value matches the reference we get from looping over. This is basically the more accurate implementation that avoids removing a key when the value was updated in the meantime.